game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/StrengthAndRollCalculator.java

```diff
@@ -0,0 +1,43 @@
+package games.strategy.triplea.delegate.power.calculator;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.attachments.UnitSupportAttachment;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.collections.IntegerMap;
+
+/** Calculates the value of the dice roll and strength. */
+abstract class StrengthAndRollCalculator {
+
+  protected final AvailableSupports friendlySupportTracker;
+  protected final AvailableSupports enemySupportTracker;
+  // Keep track of the units that are providing support and whom they are providing it to
+  @Getter private final Map<Unit, IntegerMap<Unit>> supportGiven = new HashMap<>();
+
+  StrengthAndRollCalculator(
+      final AvailableSupports friendlySupport, final AvailableSupports enemySupport) {
+    this.friendlySupportTracker = friendlySupport.filter(getRuleFilter());
+    this.enemySupportTracker = enemySupport.filter(getRuleFilter());
+  }
+
+  abstract int getValue(Unit unit);
+
+  protected abstract Predicate<UnitSupportAttachment> getRuleFilter();
+
+  protected int addSupport(final Unit unit, final AvailableSupports supportTracker) {
```

Comment:
It's a bit difficult to understand the API of this method and what is being returned exactly. Is this a functional method, or are we modifying state. If modifying state, what does the return value represent?

I think it would be easier for future maintainers to have this explicitly stated rather than being left to only reverse engineer the code.